modules/extensions-mule-support/src/main/java/org/mule/runtime/module/extension/mule/internal/loader/parser/BaseMuleSdkExtensionModelParserFactory.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.mule.internal.loader.parser;
+
+import static java.lang.String.format;
+
+import org.mule.metadata.api.TypeLoader;
+import org.mule.runtime.ast.api.ComponentAst;
+import org.mule.runtime.ast.internal.model.ExtensionModelHelper;
+import org.mule.runtime.core.api.type.catalog.ApplicationTypeLoader;
+import org.mule.runtime.extension.api.ExtensionConstants;
+import org.mule.runtime.extension.api.loader.ExtensionLoadingContext;
+import org.mule.runtime.module.extension.internal.loader.parser.ExtensionModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.ExtensionModelParserFactory;
+import org.mule.runtime.module.extension.mule.internal.loader.parser.metadata.MuleSdkExtensionModelMetadataParser;
+
+import java.util.NoSuchElementException;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+/**
+ * Base {@link ExtensionModelParserFactory} implementation for Mule SDK.
+ *
+ * Implementations need to specify how to extract some parameters needed by the parser from the extension loading context.
+ *
+ * @since 4.5.0
+ */
+public abstract class BaseMuleSdkExtensionModelParserFactory implements ExtensionModelParserFactory {
+
+  @Override
+  public ExtensionModelParser createParser(ExtensionLoadingContext context) {
+    return new MuleSdkExtensionModelParser(createMetadataParser(context),
```

Comment:
This is because there is only one implementation of `MuleSdkExtensionModelParser`, I preferred to go with composition over inheritance in this case. That is why I'm providing the strategy for parsing the metadata and for getting the top level elements which are the only differences so far between applications and extensions.

But maybe you are right in that it will be simpler using inheritance with templating. Specially considering that I'm already using templating at the factory level. I can try and see how it looks.